// appium-ios-driver/build/lib/server.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.startServer = startServer;
exports.startHttpsServer = startHttpsServer;

require("source-map-support/register");

var _logger = _interopRequireDefault(require("./logger"));

var _appiumBaseDriver = require("appium-base-driver");

var _driver = require("./driver");

var _https = _interopRequireDefault(require("https"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _requestPromise = _interopRequireDefault(require("request-promise"));

var _url = _interopRequireDefault(require("url"));

var _portfinder = _interopRequireDefault(require("portfinder"));

const pem = _bluebird.default.promisifyAll(require('pem'));

async function startServer(port, host) {
  let driver = new _driver.IosDriver();
  let routeConfiguringFunction = (0, _appiumBaseDriver.routeConfiguringFunction)(driver);
  let server = await (0, _appiumBaseDriver.server)({
    routeConfiguringFunction,
    port,
    hostname: host
  });

  _logger.default.info(`IosDriver server listening on http://${host}:${port}`);

  return server;
}

async function startHttpsServer(port, host) {
  let privateKey = await pem.createPrivateKeyAsync();
  let keys = await pem.createCertificateAsync({
    days: 1,
    selfSigned: true,
    serviceKey: privateKey.key
  });
  let pemCertificate = keys.certificate;
  let httpsPort = await _portfinder.default.getPortPromise();
  const serverOpts = {
    key: keys.serviceKey,
    cert: pemCertificate
  };

  let sslServer = _https.default.createServer(serverOpts, function requestReceived(req, res) {
    _logger.default.debug(`Received HTTPS '${req.method}' request for '${req.url}'`);

    if (req.method === 'OPTIONS') {
      res.setHeader('Access-Control-Allow-Origin', '*');
      res.setHeader('Access-Control-Allow-Methods', 'POST, GET, OPTIONS');
      res.setHeader('Access-Control-Allow-Headers', 'Content-Type');
      res.writeHead(200);
      res.end();
    } else if (req.method === 'POST') {
      let body = '';
      req.on('data', function onData(data) {
        body += data;
      });
      req.on('end', async function onEnd() {
        let httpsUrl = _url.default.parse(req.url);

        let uri = `http://${host}:${port}${httpsUrl.pathname}`;

        _logger.default.debug(`Passing '${body}' to '${uri}'`);

        let options = {
          method: 'POST',
          uri,
          body
        };
        await (0, _requestPromise.default)(options);
        res.writeHead(200);
        res.end();
      });
    }
  }).listen(httpsPort);

  return {
    sslServer,
    pemCertificate,
    httpsPort
  };
}require('source-map-support').install();
