// appium-windows-driver/build/lib/installer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.downloadWAD = downloadWAD;
exports.setupWAD = setupWAD;
exports.verifyWAD = verifyWAD;
exports.installWAD = installWAD;
exports.default = exports.WAD_GUID = exports.WAD_INSTALL_PATH = void 0;

require("source-map-support/register");

var _requestPromise = _interopRequireDefault(require("request-promise"));

var _appiumSupport = require("appium-support");

var _path = _interopRequireDefault(require("path"));

var _teen_process = require("teen_process");

var _logger = _interopRequireDefault(require("./logger"));

const WAD_VER = '1.2-RC';
const WAD_DL = `https://github.com/Microsoft/WinAppDriver/releases/download/v${WAD_VER}/WindowsApplicationDriver.msi`;
const WAD_DL_MD5 = 'dbaa9a3f7416c2b73cc5cd0e7452c8d0';
let WAD_INSTALL_PATH = process.env['ProgramFiles(x86)'] || process.env.ProgramFiles || 'C:\\Program Files';
exports.WAD_INSTALL_PATH = WAD_INSTALL_PATH;
exports.WAD_INSTALL_PATH = WAD_INSTALL_PATH = _path.default.resolve(WAD_INSTALL_PATH, 'Windows Application Driver', 'WinAppDriver.exe');
const WAD_EXE_MD5 = '50d694ebfaa622ef7e4061c1bf52efe6';
const WAD_GUID = 'DDCD58BF-37CF-4758-A15E-A60E7CF20E41';
exports.WAD_GUID = WAD_GUID;

async function downloadWAD() {
  _logger.default.info(`Opening temp file to write WinAppDriver to...`);

  let tempFile = _path.default.resolve(process.env.TEMP, 'WinAppDriver.msi');

  _logger.default.info(`Will write to ${tempFile}`);

  _logger.default.info(`Downloading ${WAD_DL}...`);

  let body = await _requestPromise.default.get({
    url: WAD_DL,
    encoding: 'binary'
  });

  _logger.default.info(`Writing binary content to ${tempFile}...`);

  await _appiumSupport.fs.writeFile(tempFile, body, {
    encoding: 'binary'
  });
  let downloadedMd5 = await _appiumSupport.fs.md5(tempFile);

  if (downloadedMd5 !== WAD_DL_MD5) {
    throw new Error(`Checksum validation error: expected ${WAD_DL_MD5} but ` + `got ${downloadedMd5}`);
  }

  return tempFile;
}

async function installWAD(msiPath) {
  _logger.default.info(`Running MSI installer...`);

  await (0, _teen_process.exec)('msiexec', ['/i', msiPath, '/qn']);
}

async function verifyWAD() {
  _logger.default.info(`You must use WinAppDriver version ${WAD_VER}`);

  _logger.default.info(`Verifying WinAppDriver version ${WAD_VER} is installed via comparing the checksum.`);

  return (await _appiumSupport.fs.exists(WAD_INSTALL_PATH)) && (await _appiumSupport.fs.md5(WAD_INSTALL_PATH)) === WAD_EXE_MD5;
}

async function isAdmin() {
  let testFilePath = _path.default.resolve('/', 'Windows', 'System32', 'wad-test.txt');

  try {
    await _appiumSupport.fs.rimraf(testFilePath);
    await _appiumSupport.fs.writeFile(testFilePath, 'foo');
    return true;
  } catch (ign) {
    return false;
  }
}

async function setupWAD() {
  if (!_appiumSupport.system.isWindows()) {
    throw new Error(`Can only download WinAppDriver on Windows!`);
  }

  if (await verifyWAD()) {
    _logger.default.info(`WinAppDriver.exe version ${WAD_VER} already exists with correct checksum, not re-downloading`);

    return;
  }

  _logger.default.info(`WinAppDriver.exe doesn't exist at the correct version ${WAD_VER}, setting up`);

  if (!(await isAdmin())) {
    throw new Error(`WARNING: You are not running as an administrator so WinAppDriver cannot be installed for you; please reinstall as admin`);
  }

  const msiPath = await downloadWAD();
  await installWAD(msiPath);

  if (!(await verifyWAD())) {
    throw new Error(`Installed version of WinAppDriver failed checksum check`);
  }
}

var _default = setupWAD;
exports.default = _default;require('source-map-support').install();
